src/client/dfuse/pil4dfs/int_dfs.c

```diff
@@ -0,0 +1,4787 @@
+			next_free_dirfd = i;
+			break;
+		}
+	}
+	if (next_free_dirfd < 0)
+		printf("WARNING> All space for dir_list are used.\n");
+
+	num_dirfd++;
+	rc = pthread_mutex_unlock(&lock_dirfd);
+	D_ASSERT(rc == 0);
+
+	if (idx < 0 && allocated)
+		D_FREE(new_obj);
+
+	return idx;
+}
+
+static int
+find_next_available_map(void)
+{
+	int i, idx = -1, rc;
+
+	rc = pthread_mutex_lock(&lock_mmap);
+	D_ASSERT(rc == 0);
+	if (next_free_map < 0) {
+		rc = pthread_mutex_unlock(&lock_mmap);
+		D_ASSERT(rc == 0);
+		return next_free_map;
+	}
+	idx = next_free_map;
+	if (next_free_map > last_map)
+		last_map = next_free_map;
+	next_free_map = -1;
+
+	for (i = idx + 1; i < MAX_MMAP_BLOCK; i++) {
+		if (mmap_list[i].addr == NULL) {
+			/* available, then update next_free_map */
+			next_free_map = i;
+			break;
+		}
+	}
+	if (next_free_map < 0)
+		printf("WARNING> All space for mmap_list are used.\n");
+
+	num_map++;
+	rc = pthread_mutex_unlock(&lock_mmap);
+	D_ASSERT(rc == 0);
+
+	return idx;
+}
+
+/* May need to support duplicated fd as duplicated dirfd too. */
+static void
+free_fd(int idx)
+{
+	int              i, rc;
+	struct file_obj *saved_obj = NULL;
+
+	rc = pthread_mutex_lock(&lock_fd);
+	D_ASSERT(rc == 0);
+	if (file_list[idx]->idx_mmap >= 0 && file_list[idx]->idx_mmap < MAX_MMAP_BLOCK) {
+		/* file_list[idx].idx_mmap >= MAX_MMAP_BLOCK means fd needs closed after munmap()*/
+		file_list[idx]->idx_mmap += MAX_MMAP_BLOCK;
+		rc = pthread_mutex_unlock(&lock_fd);
+		D_ASSERT(rc == 0);
+		return;
+	}
+
+	file_list[idx]->ref_count--;
+	if (file_list[idx]->ref_count == 0)
+		saved_obj = file_list[idx];
+	file_list[idx] = NULL;
+
+	if (idx < next_free_fd)
+		next_free_fd = idx;
+
+	if (idx == last_fd) {
+		for (i = idx - 1; i >= 0; i--) {
+			if (file_list[i]) {
+				last_fd = i;
+				break;
+			}
+		}
+	}
+
+	num_fd--;
+	rc = pthread_mutex_unlock(&lock_fd);
+	D_ASSERT(rc == 0);
+
+	if (saved_obj) {
+		rc = dfs_release(saved_obj->file);
+		memset(saved_obj, 0, sizeof(struct file_obj));
+		D_FREE(saved_obj);
+		if (rc && bLog)
+			printf("DBG> dfs_release() failed. %s\n", strerror(rc));
+	}
+}
+
+static void
+free_dirfd(int idx)
+{
+	int             i, rc;
+	struct dir_obj *saved_obj = NULL;
+
+	rc = pthread_mutex_lock(&lock_dirfd);
+	D_ASSERT(rc == 0);
+
+	dir_list[idx]->ref_count--;
+	if (dir_list[idx]->ref_count == 0)
+		saved_obj = dir_list[idx];
+	dir_list[idx] = NULL;
+
+	if (idx < next_free_dirfd)
+		next_free_dirfd = idx;
+
+	if (idx == last_dirfd) {
+		for (i = idx - 1; i >= 0; i--) {
+			if (dir_list[i]) {
+				last_dirfd = i;
+				break;
+			}
+		}
+	}
+
+	num_dirfd--;
+	rc = pthread_mutex_unlock(&lock_dirfd);
+	D_ASSERT(rc == 0);
+
+	if (saved_obj) {
+		/* free memory for path and ents. */
+		D_FREE(saved_obj->path);
+		rc = dfs_release(saved_obj->dir);
+		if (rc) {
+			printf("Warning> failed to call dfs_release(). rc = %d, %s\n",
+			       rc, strerror(rc));
+		}
+		memset(saved_obj, 0, sizeof(struct dir_obj));
```

Comment:
Again, you don't need to memset on free.

Reply:
Same as the above one.